(120) User can diagnose their problem by answering a series of questions

leeky opened the pull request:
Implement a decision tree which is read from a YAML file.

leeky committed:
(feature) QuestionSet loads questions from a YAML file

leeky committed:
(feature) Questions can be retrieved from a QuestionSet

Questions returned from `QuestionSet#find` are now `Question`
instances rather than simple hashes.

leeky committed:
(feature) User can view the first diagnosis question

leeky committed:
(feature) Question shows its answers as radio buttons

leeky committed:
(feature) After submitting an answer, user is redirected

 - If the answer has a 'next' key set, the user is redirected to the
   next question, with the id stored within the 'next' key.
 - If the answer has no 'next' key set, we redirect the user to the
   'describe repair' page.

leeky committed:
(chore) Use high_voltage to handle static content

In order to easily link to static content from the branching question
tree, there needs to be some sort of consistency.

We were serving static content from multiple places, all with uniquely
named routes, and it was all getting a bit unweildy.

Now, to add a static page, you just need to place a HAML file in
`app/views/pages`. The `page_path(...)` helper lets you link to the
individual pages.

i.e. `app/views/pages/static.html.haml` -> `page_path('static')`.

leeky committed:
(feature) After choosing an answer, the user may be shown static content

leeky committed:
(feature) Questions can also request text input

If a question has no answers, we render a textarea, for example when
requesting a description of the fault.

leeky committed:
(feature) Text entry questions redirect after submission

Previously, only questions with answers could cause a redirect. Now the
`goto` key can be set at the question level.

I've assumed that after text entry, you'll always be redirecting to
another question rather than static content. If `goto` is not set, it'll
default to redirecting the the address search form, which is the first
page AFTER the decision tree.

leeky committed:
(chore) Connect up the other pages to the decison tree

leeky committed:
(feature) User sees an error if they submit an unanswered question

If the user manages to submit the form without choosing one of the
possible answer, redisplay the form along with an error message.

leeky committed:
(chore) Update decision tree with Jason's latest changes

dgmstuart commented on spec/models/question_set_spec.rb:
```diff
@@ -0,0 +1,43 @@
+
+  describe '#find' do
+    it 'returns the requested question' do
+      store = QuestionSet.new('spec/fixtures/basic_question.yml')
```
Nice approach. Gives a lot of confidence. I thought they might be slow since they're hitting the disk, but they're 10x faster than the feature specs.

dgmstuart commented on spec/models/question_set_spec.rb:
```diff
@@ -0,0 +1,43 @@
+      }
+
+      expect(store.questions).to eql expected
+    end
```
There are no tests for how this can go wrong, but maybe we don't need to worry yet?

I guess if the file isn't YAML we'll get a YAML error, which is fine. Not sure what the implications of the structure being wrong would be (e.g. if `answers` was not an array).

leeky replied:
I've added [some tests](92b58d5ea8be928127cb27786e230be7c7f5e753) that do basic YAML validation.

dgmstuart commented on app/models/question_set.rb:
```diff
@@ -14,7 +14,7 @@ def find(id)
     end
 
-    questions.fetch(id)
+    Question.new(questions.fetch(id))
```
If we created these objects on initialization rather than on `find`, would that mean that we caught some possible mistakes in the YAML at boot time rather than at runtime? Would it be worth it?

dgmstuart commented on app/models/question_set.rb:
```diff
@@ -0,0 +1,20 @@
+class QuestionSet
+  class UnknownQuestion < StandardError; end
+
+  attr_reader :questions
```
Is this only exposed for the purposes of testing?

dgmstuart commented on spec/features/users_can_diagnose_their_issue_spec.rb:
```diff
@@ -0,0 +1,16 @@
+RSpec.feature 'Users can diagnose their issue' do
+  scenario 'viewing the first question' do
+    allow_any_instance_of(QuestionSet)
+      .to receive(:questions)
```
The method which our application code is calling is `find` - would it make more sense to stub that instead?

Is this currently relying on the fact that we know that `find` is defined in terms of `questions`?

dgmstuart commented on spec/models/question_set_spec.rb:
```diff
@@ -0,0 +1,43 @@
+require 'rails_helper'
+
+RSpec.describe QuestionSet do
+  describe '#initialize' do
```
Is this actually a test of the `#questions` method rather than intialization?

dgmstuart commented on spec/models/question_set_spec.rb:
```diff
@@ -0,0 +1,43 @@
+        },
+      }
+
+      expect(store.questions).to eql expected
```
Is this test redundant? Do we care how `QuestionSet` internally represents the data? If not then is the test of the #find method sufficient?

leeky replied:
:+1:

dgmstuart commented on spec/models/question_set_spec.rb:
```diff
@@ -0,0 +1,43 @@
+      expect(store.find('example')).to eql expected
+    end
+
+    it 'raises an error if the requestion question was not found' do
```
Typo "requestion question" 

(maybe only change if you touch this file for other reasons?)

dgmstuart commented on spec/features/users_can_diagnose_their_issue_spec.rb:
```diff
@@ -0,0 +1,16 @@
+
+RSpec.feature 'Users can diagnose their issue' do
+  scenario 'viewing the first question' do
+    allow_any_instance_of(QuestionSet)
```
`allow_any_instance_of` is intended for use with legacy code and the RSpec team recommend avoiding it if possible: https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance

Would the approach we took for `JsonApi` work here? e.g. in spec/features/residents_can_see_confirmation_of_repair_request_spec.rb

leeky replied:
We discussed a solution, and I've push a commit which switches to an `instance_double`-based stub 👍

dgmstuart commented on spec/features/users_can_diagnose_their_issue_spec.rb:
```diff
@@ -20,4 +20,56 @@
+    expect(page).to have_unchecked_field 'Absolutely!'
+  end
+
+  scenario 'choosing an answer redirects to the address search page by default' do
```
The title says "address search page" - should this be "problem description"?

dgmstuart commented on app/models/question.rb:
```diff
@@ -10,5 +10,15 @@ def initialize(question_hash)
+    if answer_hash.key?('goto')
+      Rails.application.routes.url_helpers.questions_path(answer_hash['goto'])
+    else
+      Rails.application.routes.url_helpers.describe_repair_path
```
Did you consider putting the `Rails.application.routes.url_helpers` in the initializer?

dgmstuart commented on spec/models/question_spec.rb:
```diff
@@ -32,4 +32,20 @@
+        ],
+      )
+
+      expect(question.redirect_path_for_answer('London')).to eql '/questions/next'
```
Is this now an integration test since it's testing the routing layer as well?

dgmstuart commented on app/models/question.rb:
```diff
@@ -10,5 +10,15 @@ def initialize(question_hash)
+    answer_hash = answers.detect { |answer| answer['text'] == chosen_answer }
+
+    if answer_hash.key?('goto')
+      Rails.application.routes.url_helpers.questions_path(answer_hash['goto'])
```
Is `goto` always referring to a question ID? If so is there a more descriptive name we could give it?

leeky replied:
I mainly chose it because it has the same number of characters as `text` - so it's possibly easier to read when you're editing the file as a non-developer.

leeky replied:
As discussed, I've changed `goto` to `next`.

dgmstuart commented on app/views/pages/toilet_unblock_info.html.haml:
```diff
@@ -0,0 +1,8 @@
+
+%ul
+  %li A large bin bag
+  %li A mop
```
Do we have/need something in Trello to make sure we fill these in with real content at some point?

leeky replied:
I've added a card to the backlog for these - https://trello.com/c/OyIiRYa3/146-content-for-static-pages

dgmstuart commented on db/questions.yml:
```diff
@@ -35,43 +35,155 @@ kitchen_problem:
+bathroom_lights_problem:
+  question: Is the problem with the...
+  answers:
+    - text: light fitting
```
Should we standardise on Capitalised text?

leeky replied:
:+1:

dgmstuart commented on db/questions.yml:
```diff
@@ -129,29 +244,52 @@ fire_door:
       goto: blocked_only_toilet
 
-    - text: It's cracked
+    - text: not flushing
```
Capitalise these?

dgmstuart reviewed (changes requested):
Good job

A few comments and questions

leeky committed:
(fix) Use `instance_double` instead of `allow_any_instance_of`

I couldn't quite figure out what I needed to stub in order to provide
enough of QuestionSet to allow the decision tree to be tested. Thanks to
@dgmstuart for his assistance in figuring this out!

leeky committed:
(feature) QuestionSet raises error if YAML malformed

Added some basic syntax checking of the YAML file when it is
loaded in `QuestionSet#initialize`:

 - If the YAML file cannot be parsed, raise
   `QuestionSet::BadlyFormattedYaml`
 - If the YAML file contains questions whose `next` references a
   question that does not exist, raise
   `QuestionSet::MissingQuestions`

leeky commented:
@dgmstuart This should be ready for another look.

leeky committed:
(fix) QuestionSet no longer exposes 'questions'

Due to the changes in the way `QuestionSet` is tested, we no longer
need to expose `questions` through an `attr_reader`, so remove it!

dgmstuart reviewed (approved):
LGTM - Great job!
